mlkem/native/x86_64/src/basemul.S

```diff
@@ -115,23 +104,56 @@ vmovdqa		%ymm11,(64*\off+48)*2(%rdi)
 .text
 .global MLKEM_ASM_NAMESPACE(basemul_avx2)
 MLKEM_ASM_NAMESPACE(basemul_avx2):
-mov		%rsp,%r8
+mov		%rsp,%r11
 and		$-32,%rsp
 sub		$32,%rsp
 
-lea		(_ZETAS_EXP+176)*2(%rcx),%r9
 schoolbook	0
 
-add		$32*2,%r9
 schoolbook	1
 
-add		$192*2,%r9
 schoolbook	2
 
-add		$32*2,%r9
 schoolbook	3
 
-mov		%r8,%rsp
+mov		%r11,%rsp
+ret
+
+
+.macro mulcache_compute_iter i
+vmovdqa (64*\i+16)*2(%rsi), %ymm2
+vmovdqa (64*\i+48)*2(%rsi), %ymm3
+vmovdqa (16*\i)*2(%rdx), %ymm4
+
+vpmullw %ymm2, %ymm1, %ymm5
+vpmullw %ymm3, %ymm1, %ymm6
+vpmullw %ymm5, %ymm4, %ymm5
+vpmullw %ymm6, %ymm4, %ymm6
+
+vpmulhw %ymm2, %ymm4, %ymm7
+vpmulhw %ymm3, %ymm4, %ymm8
+vpmulhw %ymm5, %ymm0, %ymm9
+vpmulhw %ymm6, %ymm0, %ymm10
+
+vpsubw %ymm9, %ymm7, %ymm7
+vpsubw %ymm10, %ymm8, %ymm8
+
+vmovdqa %ymm7, (32*\i)*2(%rdi)
+vmovdqa %ymm8, (32*\i+16)*2(%rdi)
+.endm
+
+.text
+.global MLKEM_ASM_NAMESPACE(mulcache_compute_avx2)
```

Comment:
What exactly is being calculated here? We are using half the twiddles compared to the AArch64 case, omitting negated ones -- but there is no explicit negation in the assembly. Is half of the mulcache stored negated?

Reply:
Ok yes, that does seem to be what's happening, but we should really document it.

Quite clever how the AVX2 backend arranges this. I wonder if the AArch64 backend could be similarly optimized.